doc/sphinx/how-to/how-to-enable-graphics-core22-on-a-device.md

```diff
@@ -0,0 +1,497 @@
+      usr/share/X11: X11
+    prime:
+      - usr/lib
+      - X11
+
+  wayland:
+    # Wayland support (not much cost to having this)
+    plugin: nil
+    stage-packages:
+      - libwayland-client0
+      - libwayland-cursor0
+      - libwayland-server0
+    # Mali drivers contain a build of libwayland-egl
+    prime:
+      - usr/lib
+```
+
+When you're ready, build your snap using `snapcraft`.
+
+### Troubleshooting
+There are a number of ways the vendor drivers can differ from this simplest case, requiring different solutions.
+Let's explore a few of them.
+
+#### Vendor drivers require libraries newer than core22 provides
+Some vendor driver binaries might be built against newer libraries than are provided in the 
+Ubuntu 22.04 repositories. Particularly, the core `libwayland` libraries may introduce new
+features since 22.04 that vendor binaries might require. This will manifest as missing 
+symbol errors at runtime.
+
+Generally, new version requirements for libraries can be handled by adding a snapcraft
+part building the required library version. An example of this for `libwayland` is:
+
+```yaml
+parts:
+  ...
+  wayland:
+    plugin: meson
+    meson-parameters:
+      - -Ddocumentation=false
+      - -Dtests=false
+      - --prefix=$SNAPCRAFT_PART_INSTALL/usr
+    build-packages:
+      - libxml2-dev
+      - libffi-dev
+      - libexpat1-dev
+      - pkg-config
+    override-build: |
+      snapcraftctl build
+    source: https://gitlab.freedesktop.org/wayland/wayland.git
+    source-type: git
+    source-tag: 1.21.0
+    prime:
+      - usr/lib
+```
+
+#### Vendor driver uses a $VENDOR_dri.so DRI driver
+Some vendor drivers (internally based on Mesa) retain the separate DRI driver component. If
+a vendor driver is missing its DRI component then running a program using EGL will generate
+a `libGL` warning about failure to load `$VENDOR_dri.so`, followed by either program failure
+or possibly the program continuing with CPU rendering.
+
+Any vendor DRI driver should be provided under the top-level dri/ directory. An example
+from the `imx-gpu-viv` driver is:
+
+```yaml
+parts:
+  ...
+  vendor-drivers:
+    ...
+    organize:
+      imx-gpu-viv-6.4.3.p0.0-aarch64/gpu-core/usr/lib/dri/* : graphics/dri/
+```
+
+Snaps which use the `graphics-core22` interface will set the necessary environment variables
+to point the DRI loader at the top-level `dri/` path under the path exported by the
+`graphics-core22` slot.
+
+#### Vendor drivers use libglvnd
+Some drivers might use the `libglvnd` vendor-neutral GL/EGL library. In this case the vendor
+drivers will ship an `egl_vendor.d/$SOMETHING.json` file, pointing to their vendor EGL library.
+The [mesa-core22](https://github.com/canonical/mesa-core22/blob/main/snap/snapcraft.yaml)
+`snapcraft.yaml` provides an example of this sort of driver; the `egl_vendor.d`
+directory is shipped in a top-level `usr/share/glvnd` path, under the path exported by the
+`graphics-core22` slot.
```

Comment:
This whole bit needs to change significantly, sorry.

We need to make clear a couple of things:
* The snap should provide as many of [the supported ABIs](https://mir-server.io/docs/the-graphics-core22-snap-interface#heading--lists--supported-apis) as possible.
* Those libraries don't need to be in any particular location.
* There's only a single file that the provider *does* need to have in a specific location: `bin/graphics-core22-provider-wrapper`, which gets called as `graphics-core22-provider-wrapper $SOME_BINARY $ARGS...` and needs to set up the environment so that `$SOME_BINARY` will find the libraries and then execute `$SOME_BINARY` with `$ARGS...`.

We should probably do a simple example (a libmali provider?) and then follow that up with an annotated description of the [`mesa-core22` script](https://github.com/canonical/mesa-core22/blob/main/scripts/bin/graphics-core22-provider-wrapper.in) that covers the `DRI_DRIVER_PATH` and glvnd issues.